File: src/tools/pmd-cpd/eval-repos/synthetic/csharp/SemanticDupIdentifiers.cs
/// <summary>
/// C# file with semantic duplicates - same logic with different variable names.
/// </summary>

using System;

namespace Synthetic.SemanticDuplicates
{
    public class ProfileData
    {
        public int Age { get; set; }
        public int ActivityCount { get; set; }
        public int PurchaseCount { get; set; }
        public int ReferralCount { get; set; }
        public int Id { get; set; }
        public string Name { get; set; }
        public string Email { get; set; }
        public bool Active { get; set; }
        public string CreatedAt { get; set; }
        public string LastSeen { get; set; }
    }

    public static class ScoreCalculator
    {
        public static double CalculateUserScore(ProfileData userData)
        {
            double baseValue = 100.0;
            int userAge = userData.Age;
            int userActivity = userData.ActivityCount;
            int userPurchases = userData.PurchaseCount;
            int userReferrals = userData.ReferralCount;

            if (userAge > 18) baseValue += 10.0;
            if (userAge > 30) baseValue += 5.0;
            if (userActivity > 50) baseValue += 20.0;
            if (userPurchases > 10) baseValue += 15.0;
            if (userReferrals > 5) baseValue += 25.0;

            return Math.Min(baseValue, 200.0);
        }

        public static double CalculateCustomerRating(ProfileData customerInfo)
        {
            double startingScore = 100.0;
            int customerYears = customerInfo.Age;
            int customerInteractions = customerInfo.ActivityCount;
            int customerOrders = customerInfo.PurchaseCount;
            int customerInvites = customerInfo.ReferralCount;

            if (customerYears > 18) startingScore += 10.0;
            if (customerYears > 30) startingScore += 5.0;
            if (customerInteractions > 50) startingScore += 20.0;
            if (customerOrders > 10) startingScore += 15.0;
            if (customerInvites > 5) startingScore += 25.0;

            return Math.Min(startingScore, 200.0);
        }
    }

    public class ProcessedRecord
    {
        public int Identifier { get; set; }
        public string FullName { get; set; }
        public string EmailAddress { get; set; }
        public bool IsActive { get; set; }
        public string RegistrationDate { get; set; }
        public string LastLogin { get; set; }
    }

    public class ProcessedMember
    {
        public int MemberId { get; set; }
        public string DisplayName { get; set; }
        public string ContactEmail { get; set; }
        public bool AccountStatus { get; set; }
        public string SignupDate { get; set; }
        public string RecentActivity { get; set; }
    }

    public static class RecordProcessor
    {
        public static ProcessedRecord ProcessUserRecord(ProfileData record)
        {
            return new ProcessedRecord
            {
                Identifier = record.Id,
                FullName = (record.Name ?? "").Trim(),
                EmailAddress = (record.Email ?? "").ToLower().Trim(),
                IsActive = record.Active,
                RegistrationDate = record.CreatedAt ?? "",
                LastLogin = record.LastSeen ?? ""
            };
        }

        public static ProcessedMember ProcessMemberEntry(ProfileData entry)
        {
            return new ProcessedMember
            {
                MemberId = entry.Id,
                DisplayName = (entry.Name ?? "").Trim(),
                ContactEmail = (entry.Email ?? "").ToLower().Trim(),
                AccountStatus = entry.Active,
                SignupDate = entry.CreatedAt ?? "",
                RecentActivity = entry.LastSeen ?? ""
            };
        }
    }
}

File: src/tools/pmd-cpd/eval-repos/synthetic/typescript/semantic_dup_identifiers.ts
/**
 * TypeScript file with semantic duplicates - same logic with different variable names.
 */

interface ProfileData {
    age?: number;
    activityCount?: number;
    purchaseCount?: number;
    referralCount?: number;
    id?: number;
    name?: string;
    email?: string;
    active?: boolean;
    createdAt?: string;
    lastSeen?: string;
}

function calculateUserScore(userData: ProfileData): number {
    let baseValue = 100.0;
    const userAge = userData.age || 0;
    const userActivity = userData.activityCount || 0;
    const userPurchases = userData.purchaseCount || 0;
    const userReferrals = userData.referralCount || 0;

    if (userAge > 18) baseValue += 10.0;
    if (userAge > 30) baseValue += 5.0;
    if (userActivity > 50) baseValue += 20.0;
    if (userPurchases > 10) baseValue += 15.0;
    if (userReferrals > 5) baseValue += 25.0;

    return Math.min(baseValue, 200.0);
}

function calculateCustomerRating(customerInfo: ProfileData): number {
    let startingScore = 100.0;
    const customerYears = customerInfo.age || 0;
    const customerInteractions = customerInfo.activityCount || 0;
    const customerOrders = customerInfo.purchaseCount || 0;
    const customerInvites = customerInfo.referralCount || 0;

    if (customerYears > 18) startingScore += 10.0;
    if (customerYears > 30) startingScore += 5.0;
    if (customerInteractions > 50) startingScore += 20.0;
    if (customerOrders > 10) startingScore += 15.0;
    if (customerInvites > 5) startingScore += 25.0;

    return Math.min(startingScore, 200.0);
}

interface ProcessedRecord {
    identifier?: number;
    fullName?: string;
    emailAddress?: string;
    isActive?: boolean;
    registrationDate?: string;
    lastLogin?: string;
    memberId?: number;
    displayName?: string;
    contactEmail?: string;
    accountStatus?: boolean;
    signupDate?: string;
    recentActivity?: string;
}

function processUserRecord(record: ProfileData): ProcessedRecord {
    return {
        identifier: record.id || 0,
        fullName: (record.name || "").trim(),
        emailAddress: (record.email || "").toLowerCase().trim(),
        isActive: record.active !== false,
        registrationDate: record.createdAt || "",
        lastLogin: record.lastSeen || "",
    };
}

function processMemberEntry(entry: ProfileData): ProcessedRecord {
    return {
        memberId: entry.id || 0,
        displayName: (entry.name || "").trim(),
        contactEmail: (entry.email || "").toLowerCase().trim(),
        accountStatus: entry.active !== false,
        signupDate: entry.createdAt || "",
        recentActivity: entry.lastSeen || "",
    };
}

export { ProfileData, ProcessedRecord, calculateUserScore, calculateCustomerRating, processUserRecord, processMemberEntry };
